dask/order.py

```diff
@@ -115,786 +106,434 @@ def order(
+    sort_keys = {
+        # Constructing those tuples is relatively expensive if done during a
+        # sorting operation. Therefore, compute it once and cache it
+        x: (
+            len(roots_connected[x]),
+            total_dependencies[x],
+            -max(len(dependents[k]) for k in roots_connected[x]),
             StrComparable(x),
         )
-
-    root_total_dependencies = total_dependencies[list(root_nodes)[0]]
-    # Computing this for all keys can sometimes be relatively expensive :(
-    partition_keys = {
-        key: (
-            (root_total_dependencies - total_dependencies[key] + 1)
-            * (total_dependents - min_heights)
-        )
-        for key, (
-            total_dependents,
-            _,
-            _,
-            min_heights,
-            _,
-        ) in metrics.items()
+        for x in dsk
     }
+    sort_key = sort_keys.__getitem__
+
+    def add_to_result(item: Key) -> None:
+        nonlocal crit_path_counter
+        # Earlier versions recursed into this method but this could cause
+        # recursion depth errors
+        next_items = [item]
+        nonlocal i
+        while next_items:
+            item = next_items.pop()
+            assert not num_needed[item]
+            linear_hull.discard(item)
+            if item in result:
+                continue
+            if return_stats:
+                result[item] = Order(i, crit_path_counter - _crit_path_counter_offset)
+            else:
+                result[item] = i
+            i += 1
+            # Note: This is a `set` and therefore this introduces a certain
+            # randomness. However, this randomness should not have any impact on
+            # the final result since the `process_runnable` should produce
+            # equivalent results regardless of the order in which runnable is
+            # populated (not identical but equivalent)
+            for dep in dependents[item]:
+                num_needed[dep] -= 1
+                if not num_needed[dep]:
+                    if len(dependents[item]) == 1:
+                        next_items.append(dep)
+                    else:
+                        runnable.append(dep)
+
+    def _with_offset(func: Callable[..., None]) -> Callable[..., None]:
+        # This decorator is only used to reduce indentation levels. The offset
+        # is purely cosmetical and used for some visualizations and I haven't
+        # settled on how to implement this best so I didn't want to have large
+        # indentations that make things harder to read
+        nonlocal _crit_path_counter_offset
+
+        def wrapper(*args: Any, **kwargs: Any) -> None:
+            nonlocal _crit_path_counter_offset
+            _crit_path_counter_offset = 0.5
+            try:
+                func(*args, **kwargs)
+            finally:
+                _crit_path_counter_offset = 0
+
+        return wrapper
+
+    @_with_offset
+    def process_runnables() -> None:
+        while runnable:
```

Comment:
This was added after the latest benchmarks ran. While all test are green and some are performing better, I will rerun the coiled benchmarks suite to confirm this isn't causing problems

Reply:
Is there an update on this?

Reply:
https://github.com/coiled/benchmarks/actions/runs/7165432161

Results look pretty much as before. I see again a regression in the test_join cases that I couldn't reproduce locally but I am rerunning the benchmarks for those tests specifically to double check